src/aleph/http/http2.clj

```diff
@@ -49,102 +91,356 @@
-   connection-related headers. Throws on nil header values. Throws if
+   Header names are converted to lower-case AsciiStrings. Values are left as-is,
+   but it's up to the user to ensure they are valid. For more info, see
+   https://www.rfc-editor.org/rfc/rfc9113.html#section-8.2.1
+
+   Respects HTTP/2 rules. All headers will be made lower-case. Throws on
+   invalid connection-related headers. Throws on nil header values. Throws if
    `transfer-encoding` is present, but is not equal to 'trailers'."
   [^Http2Headers h2-headers ^String header-name header-value]
-  (log/debug "Adding HTTP header" header-name ": " header-value)
 
+  ;; Also checked by Netty, but we want to avoid work, so we check too
   (if (nil? header-name)
-    (illegal-arg "Header name cannot be nil")
+    (throw-illegal-arg-ex "Header name cannot be nil")
 
     ;; Technically, Ring requires lower-cased headers, but there's no guarantee, and this is
     ;; probably faster, since most users won't be caching.
     (let [header-name (if-some [cached (.get cached-header-names header-name)]
                         cached
-                        (let [lower-cased (str/lower-case (name header-name))]
+                        (let [lower-cased (-> header-name (name) (AsciiString.) (.toLowerCase))]
                           (.put cached-header-names header-name lower-cased)
                           lower-cased))]
+      (log/trace (str "Adding HTTP header: " header-name ": " header-value))
+
       (cond
-        (nil? header-value)
-        (illegal-arg (str "Invalid nil value for header '" header-name "'"))
+        ;; Will be checked by Netty
+        ;;(nil? header-value)
+        ;;(illegal-arg (str "Invalid nil value for header '" header-name "'"))
 
         (invalid-headers header-name)
-        (illegal-arg (str "Forbidden HTTP/2 header: \"" header-name "\""))
+        (throw-illegal-arg-ex (str "Forbidden HTTP/2 header: \"" header-name "\""))
 
         ;; See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
-        (and (.equals "transfer-encoding" header-name)
+        (and (.equals te-header-name header-name)
              (not (.equals "trailers" header-value)))
-        (illegal-arg "Invalid value for 'transfer-encoding' header. Only 'trailers' is allowed.")
+        (throw-illegal-arg-ex "Invalid value for 'transfer-encoding' header. Only 'trailers' is allowed.")
 
         (sequential? header-value)
         (.add h2-headers ^CharSequence header-name ^Iterable header-value)
 
         :else
         (.add h2-headers ^CharSequence header-name ^Object header-value)))))
 
-(defn- ring-map->netty-http2-headers
-  "Builds a Netty Http2Headers object from a Ring map."
+
+(defn parse-status
+  "Parses the HTTP status of a ring response map. Returns an HttpResponseStatus"
+  ^HttpResponseStatus
+  [status]
+  (cond
+    (or (instance? Long status)
+        (instance? Integer status)
+        (instance? Short status))
```

Comment:
Could we rely `java.lang.Number` instead?

Reply:
Maybe. Number would include Bytes, Atomic*s, and Big*s, which we don't want, but that's pretty unlikely, honestly. 

Profiling suggests it's all too fast to tell the difference.

Reply:
All good then!